public/lessons/react/what_is_react.md

```diff
@@ -0,0 +1,178 @@
+- **DOM** stands for *Document-Object-Model*, and is managed by the browser.
+- What this means in practice is that React keeps a copy of the web page structure and state in **memory**.
+- This copy is accessible to JavaScript and React makes **changes** to the **real** DOM all at once instead of one at a time.
+
+# Virtual DOM 2
+
+Imagine an auction application with several parts.
+
+- The items on auction right now with current prices.
+- Your current bid, and your maximum bid ceiling.
+- A SOLD label applied to an item that has closed.
+- The items that will come up for auction soon.
+- Recently sold items and their closing prices.
+
+# Virtual DOM 3
+
+In this application there are many changes that can happen at once.
+
+- The items available can increase or decrease.
+- Your current bid can increase in response to another bidder.
+- The SOLD label can be applied or maybe removed.
+- New items can appear with their opening prices.
+- Recently sold items can appear with prices.
+
+# Virtual DOM 4
+
+React makes managing a complicated page like this easier by determining what the page needs to look like at any given rendered **frame**.
+
+# Virtual DOM 5
+
+For every given frame
+
+- Any changes within that time slot are recorded.
+- Interdependencies between components are calculated.
+- The next state of the interface is generated.
+- The frame is rendered to the browser.
+
+# Virtual DOM 6
+
+This lets you as a programmer
+
+- Declare the page you want.
+- Decide how to handle change when it occurs.
+- Allow React to build the page for you when changes happen.
+
+# Declarative Intent 1
+
+- React lets you to **declare** what you want the page to be.
+
+        @@@javascript
+        ReactDOM.render(React.createElement(
+          'h1',
+          null,
+          'Hello, React!'
+        ), document.getElementById('root'));
+
+# Declarative Intent
+
+Given a `<root>` element exists the result is:
+
+```
+<root>
+  <h1>Hello, React!</h1>
+</root>
+```
+
+# Declarative Intent
+
+- Something a little more complicated
+
+        @@@javascript
+        const element = React.createClass({
+          render: function () {
+            return React.createElement(
+              "div",
+              null,
+              React.createElement(
+                "form",
+                {
+                  id: "my-form",
+                  onSubmit: this.handleSubmit
+                },
+                React.createElement(
+                  "input",
+                  {
+                    id: "create",
+                    type: "text",
+                    placeholder: "something"
+                  }
+                )
+              )
+            );
+          }
+        });
+        const container = document.getElementById('root');
+        ReactDOM.render(element, container);
+
+const container = document.getElementById('root');
+ReactDOM.render(element, container);
```

Comment:
This is a lot of code for one slide... maybe break it up by extracting a few methods, e.g. `renderInput` and `renderForm`.

Also, the different meanings of `render` need exposition: this class has a `render` method, but `ReactDOM` *also* has a different `render` method that calls ours.

Do we want to put this off until after the next few slides, or does it make sense to show them the whole thing and say "don't worry if you don't get it all yet; we'll come back to this a bit later"?